File: sdk/csharp/CLI/Commands/DevelopmentCommands.cs
using System;
using System.CommandLine;
using System.CommandLine.Invocation;
using System.IO;
using System.Threading.Tasks;
using System.Net;
using System.Text;
using System.Collections.Generic;
using System.Linq;

namespace TuskLang.CLI.Commands
{
    /// <summary>
    /// Development commands for TuskLang CLI
    /// </summary>
    public static class DevelopmentCommands
    {
        public static void AddCommands(RootCommand rootCommand)
        {
            var serveCommand = new Command("serve", "Start development server")
            {
                new Argument<int>("port", () => 8080, "Port number"),
                new Option<string>("--host", () => "localhost", "Host address"),
                new Option<bool>("--https", "Enable HTTPS"),
                Handler = CommandHandler.Create<int, string, bool>(StartDevelopmentServer)
            };

            var compileCommand = new Command("compile", "Compile .tsk file")
            {
                new Argument<string>("file", "TSK file to compile"),
                new Option<string>("--output", "Output file path"),
                new Option<bool>("--optimize", "Enable optimization"),
                Handler = CommandHandler.Create<string, string, bool>(CompileTskFile)
            };

            var optimizeCommand = new Command("optimize", "Optimize .tsk file for production")
            {
                new Argument<string>("file", "TSK file to optimize"),
                new Option<string>("--output", "Output file path"),
                new Option<bool>("--minify", "Minify output"),
                Handler = CommandHandler.Create<string, string, bool>(OptimizeTskFile)
            };

            rootCommand.AddCommand(serveCommand);
            rootCommand.AddCommand(compileCommand);
            rootCommand.AddCommand(optimizeCommand);
        }

        private static async Task<int> StartDevelopmentServer(int port, string host, bool https)
        {
            try
            {
                GlobalOptions.WriteProcessing($"Starting development server on {host}:{port}");

                var config = await LoadConfiguration();
                var serverConfig = GetServerConfiguration(config);

                var listener = new HttpListener();
                var prefix = $"http{(https ? "s" : "")}://{host}:{port}/";
                listener.Prefixes.Add(prefix);
                listener.Start();

                GlobalOptions.WriteSuccess($"Development server started at {prefix}");
                GlobalOptions.WriteLine("Press Ctrl+C to stop the server");

                // Handle requests
                while (listener.IsListening)
                {
                    try
                    {
                        var context = await listener.GetContextAsync();
                        _ = Task.Run(() => HandleRequest(context, serverConfig));
                    }
                    catch (Exception ex)
                    {
                        if (listener.IsListening)
                        {
                            GlobalOptions.WriteError($"Request handling error: {ex.Message}");
                        }
                    }
                }

                return 0;
            }
            catch (Exception ex)
            {
                GlobalOptions.WriteError($"Failed to start development server: {ex.Message}");
                return 1;
            }
        }

        private static async Task HandleRequest(HttpListenerContext context, Dictionary<string, object> serverConfig)
        {
            try
            {
                var request = context.Request;
                var response = context.Response;

                GlobalOptions.WriteLine($"{request.HttpMethod} {request.Url.PathAndQuery}");

                var path = request.Url.AbsolutePath;
                var responseContent = "";

                // Route requests
                switch (path)
                {
                    case "/":
                        responseContent = GenerateHomePage(serverConfig);
                        response.ContentType = "text/html";
                        break;

                    case "/api/status":
                        responseContent = GenerateStatusResponse(serverConfig);
                        response.ContentType = "application/json";
                        break;

                    case "/api/config":
                        responseContent = GenerateConfigResponse(serverConfig);
                        response.ContentType = "application/json";
                        break;

                    default:
                        response.StatusCode = 404;
                        responseContent = Generate404Page();
                        response.ContentType = "text/html";
                        break;
                }

                var buffer = Encoding.UTF8.GetBytes(responseContent);
                response.ContentLength64 = buffer.Length;
                await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
                response.Close();
            }
            catch (Exception ex)
            {
                GlobalOptions.WriteError($"Request handling error: {ex.Message}");
            }
        }

        private static string GenerateHomePage(Dictionary<string, object> serverConfig)
        {
            return $@"
<!DOCTYPE html>
<html>
<head>
    <title>TuskLang Development Server</title>
    <style>
        body {{ font-family: Arial, sans-serif; margin: 40px; }}
        .container {{ max-width: 800px; margin: 0 auto; }}
        .header {{ background: #f0f0f0; padding: 20px; border-radius: 5px; }}
        .endpoint {{ background: #f9f9f9; padding: 10px; margin: 10px 0; border-left: 4px solid #007acc; }}
    </style>
</head>
<body>
    <div class='container'>
        <div class='header'>
            <h1>🐘 TuskLang Development Server</h1>
            <p>Strong. Secure. Scalable.</p>
        </div>
        
        <h2>Available Endpoints</h2>
        <div class='endpoint'>
            <strong>GET /api/status</strong> - Server status and health check
        </div>
        <div class='endpoint'>
            <strong>GET /api/config</strong> - Current configuration
        </div>
        
        <h2>Server Information</h2>
        <p><strong>Version:</strong> {serverConfig.GetValueOrDefault("version", "2.0.0")}</p>
        <p><strong>Environment:</strong> {serverConfig.GetValueOrDefault("environment", "development")}</p>
        <p><strong>Started:</strong> {DateTime.Now:yyyy-MM-dd HH:mm:ss}</p>
    </div>
</body>
</html>";
        }

        private static string GenerateStatusResponse(Dictionary<string, object> serverConfig)
        {
            return System.Text.Json.JsonSerializer.Serialize(new
            {
                status = "running",
                version = serverConfig.GetValueOrDefault("version", "2.0.0"),
                environment = serverConfig.GetValueOrDefault("environment", "development"),
                uptime = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss"),
                timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
            });
        }

        private static string GenerateConfigResponse(Dictionary<string, object> serverConfig)
        {
            return System.Text.Json.JsonSerializer.Serialize(serverConfig);
        }

        private static string Generate404Page()
        {
            return @"
<!DOCTYPE html>
<html>
<head>
    <title>404 - Not Found</title>
    <style>
        body { font-family: Arial, sans-serif; text-align: center; margin-top: 100px; }
    </style>
</head>
<body>
    <h1>404 - Page Not Found</h1>
    <p>The requested resource was not found on this server.</p>
    <a href='/'>Return to Home</a>
</body>
</html>";
        }

        private static async Task<int> CompileTskFile(string file, string output, bool optimize)
        {
            try
            {
                if (!File.Exists(file))
                {
                    GlobalOptions.WriteError($"File not found: {file}");
                    return 3;
                }

                GlobalOptions.WriteProcessing($"Compiling TSK file: {file}");

                var content = await File.ReadAllTextAsync(file);
                var tsk = TSK.FromString(content);

                // Compile to optimized format
                var compiledContent = CompileTskContent(tsk, optimize);

                // Determine output file
                var outputFile = output ?? Path.ChangeExtension(file, ".tskc");
                await File.WriteAllTextAsync(outputFile, compiledContent);

                GlobalOptions.WriteSuccess($"TSK file compiled to: {outputFile}");

                if (GlobalOptions.Verbose)
                {
                    var originalSize = new FileInfo(file).Length;
                    var compiledSize = new FileInfo(outputFile).Length;
                    var compression = ((double)(originalSize - compiledSize) / originalSize) * 100;
                    GlobalOptions.WriteLine($"Original size: {originalSize} bytes");
                    GlobalOptions.WriteLine($"Compiled size: {compiledSize} bytes");
                    GlobalOptions.WriteLine($"Compression: {compression:F1}%");
                }

                return 0;
            }
            catch (Exception ex)
            {
                GlobalOptions.WriteError($"Compilation failed: {ex.Message}");
                return 1;
            }
        }

        private static async Task<int> OptimizeTskFile(string file, string output, bool minify)
        {
            try
            {
                if (!File.Exists(file))
                {
                    GlobalOptions.WriteError($"File not found: {file}");
                    return 3;
                }

                GlobalOptions.WriteProcessing($"Optimizing TSK file: {file}");

                var content = await File.ReadAllTextAsync(file);
                var tsk = TSK.FromString(content);

                // Optimize the content
                var optimizedContent = OptimizeTskContent(tsk, minify);

                // Determine output file
                var outputFile = output ?? Path.ChangeExtension(file, ".optimized.tsk");
                await File.WriteAllTextAsync(outputFile, optimizedContent);

                GlobalOptions.WriteSuccess($"TSK file optimized to: {outputFile}");

                if (GlobalOptions.Verbose)
                {
                    var originalSize = new FileInfo(file).Length;
                    var optimizedSize = new FileInfo(outputFile).Length;
                    var improvement = ((double)(originalSize - optimizedSize) / originalSize) * 100;
                    GlobalOptions.WriteLine($"Original size: {originalSize} bytes");
                    GlobalOptions.WriteLine($"Optimized size: {optimizedSize} bytes");
                    GlobalOptions.WriteLine($"Size improvement: {improvement:F1}%");
                }

                return 0;
            }
            catch (Exception ex)
            {
                GlobalOptions.WriteError($"Optimization failed: {ex.Message}");
                return 1;
            }
        }

        private static string CompileTskContent(TSK tsk, bool optimize)
        {
            var data = tsk.ToDictionary();
            var compiled = new StringBuilder();

            compiled.AppendLine("# TuskLang Compiled Configuration");
            compiled.AppendLine($"# Compiled: {DateTime.Now:yyyy-MM-dd HH:mm:ss}");
            compiled.AppendLine($"# Optimized: {optimize}");
            compiled.AppendLine();

            foreach (var section in data)
            {
                compiled.AppendLine($"[{section.Key}]");

                if (section.Value is Dictionary<string, object> sectionData)
                {
                    foreach (var kvp in sectionData)
                    {
                        var value = FormatValueForCompilation(kvp.Value, optimize);
                        compiled.AppendLine($"{kvp.Key} = {value}");
                    }
                }
                else
                {
                    compiled.AppendLine($"value = {FormatValueForCompilation(section.Value, optimize)}");
                }

                compiled.AppendLine();
            }

            return compiled.ToString();
        }

        private static string OptimizeTskContent(TSK tsk, bool minify)
        {
            var data = tsk.ToDictionary();
            var optimized = new StringBuilder();

            if (!minify)
            {
                optimized.AppendLine("# TuskLang Optimized Configuration");
                optimized.AppendLine($"# Optimized: {DateTime.Now:yyyy-MM-dd HH:mm:ss}");
                optimized.AppendLine();
            }

            foreach (var section in data)
            {
                if (!minify) optimized.AppendLine($"[{section.Key}]");
                else optimized.Append($"[{section.Key}]");

                if (section.Value is Dictionary<string, object> sectionData)
                {
                    foreach (var kvp in sectionData)
                    {
                        var value = FormatValueForOptimization(kvp.Value, minify);
                        if (minify)
                            optimized.Append($"{kvp.Key}={value}");
                        else
                            optimized.AppendLine($"{kvp.Key} = {value}");
                    }
                }
                else
                {
                    var value = FormatValueForOptimization(section.Value, minify);
                    if (minify)
                        optimized.Append($"value={value}");
                    else
                        optimized.AppendLine($"value = {value}");
                }

                if (!minify) optimized.AppendLine();
            }

            return optimized.ToString();
        }

        private static string FormatValueForCompilation(object value, bool optimize)
        {
            if (value == null) return "null";
            if (value is bool b) return b.ToString().ToLower();
            if (value is string str)
            {
                if (optimize && str.Length > 100)
                {
                    // Truncate long strings in optimized mode
                    return $"\"{str.Substring(0, 100)}...\"";
                }
                return $"\"{str.Replace("\"", "\\\"")}\"";
            }
            if (value is int || value is long || value is double || value is float) return value.ToString();
            if (value is Dictionary<string, object> dict)
            {
                var pairs = dict.Select(kvp => $"\"{kvp.Key}\" = {FormatValueForCompilation(kvp.Value, optimize)}");
                return $"{{{string.Join(", ", pairs)}}}";
            }
            if (value is IEnumerable<object> enumerable)
            {
                var items = enumerable.Select(item => FormatValueForCompilation(item, optimize));
                return $"[{string.Join(", ", items)}]";
            }
            return $"\"{value}\"";
        }

        private static string FormatValueForOptimization(object value, bool minify)
        {
            if (value == null) return "null";
            if (value is bool b) return b.ToString().ToLower();
            if (value is string str)
            {
                if (minify)
                {
                    // Remove unnecessary quotes for simple strings
                    if (str.All(c => char.IsLetterOrDigit(c) || c == '_' || c == '-'))
                        return str;
                }
                return $"\"{str.Replace("\"", "\\\"")}\"";
            }
            if (value is int || value is long || value is double || value is float) return value.ToString();
            if (value is Dictionary<string, object> dict)
            {
                var pairs = dict.Select(kvp => $"\"{kvp.Key}\"={FormatValueForOptimization(kvp.Value, minify)}");
                return $"{{{string.Join(",", pairs)}}}";
            }
            if (value is IEnumerable<object> enumerable)
            {
                var items = enumerable.Select(item => FormatValueForOptimization(item, minify));
                return $"[{string.Join(",", items)}]";
            }
            return $"\"{value}\"";
        }

        private static async Task<Dictionary<string, object>> LoadConfiguration()
        {
            var config = new PeanutConfig();
            return await config.LoadAsync();
        }

        private static Dictionary<string, object> GetServerConfiguration(Dictionary<string, object> config)
        {
            return new Dictionary<string, object>
            {
                ["version"] = config.GetValueOrDefault("version", "2.0.0"),
                ["environment"] = config.GetValueOrDefault("environment", "development"),
                ["server"] = config.GetValueOrDefault("server", new Dictionary<string, object>()),
                ["features"] = config.GetValueOrDefault("features", new Dictionary<string, object>())
            };
        }
    }
} 

File: sdk/java/src/main/java/org/tusklang/cli/DevelopmentCommands.java
package org.tusklang.cli;

import org.apache.commons.cli.*;
import org.tusklang.TuskLangEnhanced;
import org.tusklang.TuskLangParser;
import java.io.*;
import java.nio.file.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.Executors;
import java.util.concurrent.ExecutorService;
import java.util.Map;

/**
 * Development Commands Implementation
 * 
 * Commands:
 * - tsk serve [port]                - Start development server
 * - tsk compile <file>              - Compile .tsk file
 * - tsk optimize <file>             - Optimize .tsk file
 */
public class DevelopmentCommands {
    
    /**
     * tsk serve [port] - Start development server
     */
    public static boolean handleServe(String[] args, CommandLine globalCmd) {
        int port = 8080; // Default port
        
        if (args.length > 0) {
            try {
                port = Integer.parseInt(args[0]);
            } catch (NumberFormatException e) {
                System.err.println("Error: Invalid port number: " + args[0]);
                return false;
            }
        }
        
        try {
            ServerSocket serverSocket = new ServerSocket(port);
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"started\",");
                System.out.println("  \"port\": " + port + ",");
                System.out.println("  \"url\": \"http://localhost:" + port + "\"");
                System.out.println("}");
            } else {
                System.out.println("🚀 TuskLang Development Server started");
                System.out.println("📍 Port: " + port);
                System.out.println("📍 URL: http://localhost:" + port);
                System.out.println("📍 Press Ctrl+C to stop");
            }
            
            ExecutorService executor = Executors.newCachedThreadPool();
            
            while (true) {
                try {
                    Socket clientSocket = serverSocket.accept();
                    executor.submit(() -> handleClient(clientSocket, globalCmd));
                } catch (IOException e) {
                    if (!serverSocket.isClosed()) {
                        System.err.println("Error accepting connection: " + e.getMessage());
                    }
                    break;
                }
            }
            
            executor.shutdown();
            serverSocket.close();
            return true;
            
        } catch (IOException e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Failed to start server: " + e.getMessage());
            }
            return false;
        }
    }
    
    /**
     * tsk compile <file> - Compile .tsk file
     */
    public static boolean handleCompile(String[] args, CommandLine globalCmd) {
        if (args.length == 0) {
            System.err.println("Error: compile command requires a file path");
            return false;
        }
        
        String inputFile = args[0];
        
        try {
            if (!Files.exists(Paths.get(inputFile))) {
                System.err.println("Error: File not found: " + inputFile);
                return false;
            }
            
            // Parse the TSK file
            TuskLangEnhanced parser = new TuskLangEnhanced();
            Map<String, Object> config = parser.parseFile(inputFile);
            
            // Generate output filename
            String outputFile = inputFile;
            if (inputFile.endsWith(".tsk")) {
                outputFile = inputFile.substring(0, inputFile.length() - 4) + ".compiled.tsk";
            } else {
                outputFile = inputFile + ".compiled";
            }
            
            // Write compiled output
            String compiledContent = generateCompiledContent(config, inputFile);
            Files.write(Paths.get(outputFile), compiledContent.getBytes());
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"success\",");
                System.out.println("  \"input_file\": \"" + inputFile + "\",");
                System.out.println("  \"output_file\": \"" + outputFile + "\",");
                System.out.println("  \"size_bytes\": " + Files.size(Paths.get(outputFile)));
                System.out.println("}");
            } else {
                System.out.println("✅ File compiled successfully");
                System.out.println("📍 Input: " + inputFile);
                System.out.println("📍 Output: " + outputFile);
                System.out.println("📍 Size: " + Files.size(Paths.get(outputFile)) + " bytes");
            }
            
            return true;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Compilation failed: " + e.getMessage());
            }
            return false;
        }
    }
    
    /**
     * tsk optimize <file> - Optimize .tsk file
     */
    public static boolean handleOptimize(String[] args, CommandLine globalCmd) {
        if (args.length == 0) {
            System.err.println("Error: optimize command requires a file path");
            return false;
        }
        
        String inputFile = args[0];
        
        try {
            if (!Files.exists(Paths.get(inputFile))) {
                System.err.println("Error: File not found: " + inputFile);
                return false;
            }
            
            // Parse the TSK file
            TuskLangEnhanced parser = new TuskLangEnhanced();
            Map<String, Object> config = parser.parseFile(inputFile);
            
            // Apply optimizations
            Map<String, Object> optimized = optimizeConfiguration(config);
            
            // Generate output filename
            String outputFile = inputFile;
            if (inputFile.endsWith(".tsk")) {
                outputFile = inputFile.substring(0, inputFile.length() - 4) + ".optimized.tsk";
            } else {
                outputFile = inputFile + ".optimized";
            }
            
            // Write optimized output
            String optimizedContent = generateOptimizedContent(optimized, inputFile);
            Files.write(Paths.get(outputFile), optimizedContent.getBytes());
            
            long originalSize = Files.size(Paths.get(inputFile));
            long optimizedSize = Files.size(Paths.get(outputFile));
            double improvement = ((double)(originalSize - optimizedSize) / originalSize) * 100;
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"success\",");
                System.out.println("  \"input_file\": \"" + inputFile + "\",");
                System.out.println("  \"output_file\": \"" + outputFile + "\",");
                System.out.println("  \"original_size\": " + originalSize + ",");
                System.out.println("  \"optimized_size\": " + optimizedSize + ",");
                System.out.println("  \"improvement_percent\": " + String.format("%.1f", improvement));
                System.out.println("}");
            } else {
                System.out.println("✅ File optimized successfully");
                System.out.println("📍 Input: " + inputFile);
                System.out.println("📍 Output: " + outputFile);
                System.out.println("📍 Original size: " + originalSize + " bytes");
                System.out.println("📍 Optimized size: " + optimizedSize + " bytes");
                System.out.println("📍 Improvement: " + String.format("%.1f", improvement) + "%");
            }
            
            return true;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Optimization failed: " + e.getMessage());
            }
            return false;
        }
    }
    
    // Helper methods
    
    private static void handleClient(Socket clientSocket, CommandLine globalCmd) {
        try (BufferedReader in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
             PrintWriter out = new PrintWriter(clientSocket.getOutputStream(), true)) {
            
            // Read HTTP request
            String requestLine = in.readLine();
            if (requestLine == null) return;
            
            String[] parts = requestLine.split(" ");
            if (parts.length < 2) return;
            
            String method = parts[0];
            String path = parts[1];
            
            // Handle different paths
            if ("GET".equals(method)) {
                if ("/".equals(path)) {
                    sendResponse(out, "200 OK", "text/html", generateIndexHtml());
                } else if ("/health".equals(path)) {
                    sendResponse(out, "200 OK", "application/json", "{\"status\":\"healthy\"}");
                } else if (path.startsWith("/config/")) {
                    handleConfigRequest(out, path.substring(8), globalCmd);
                } else {
                    sendResponse(out, "404 Not Found", "text/plain", "Not Found");
                }
            } else {
                sendResponse(out, "405 Method Not Allowed", "text/plain", "Method Not Allowed");
            }
            
        } catch (IOException e) {
            // Ignore client errors
        } finally {
            try {
                clientSocket.close();
            } catch (IOException e) {
                // Ignore
            }
        }
    }
    
    private static void sendResponse(PrintWriter out, String status, String contentType, String body) {
        out.println("HTTP/1.1 " + status);
        out.println("Content-Type: " + contentType + "; charset=UTF-8");
        out.println("Content-Length: " + body.getBytes().length);
        out.println("Access-Control-Allow-Origin: *");
        out.println();
        out.println(body);
    }
    
    private static String generateIndexHtml() {
        return """
            <!DOCTYPE html>
            <html>
            <head>
                <title>TuskLang Development Server</title>
                <style>
                    body { font-family: Arial, sans-serif; margin: 40px; }
                    .endpoint { background: #f5f5f5; padding: 10px; margin: 10px 0; border-radius: 5px; }
                    .method { color: #0066cc; font-weight: bold; }
                </style>
            </head>
            <body>
                <h1>🐘 TuskLang Development Server</h1>
                <p>Welcome to the TuskLang development server!</p>
                
                <h2>Available Endpoints:</h2>
                <div class="endpoint">
                    <span class="method">GET</span> / - This page
                </div>
                <div class="endpoint">
                    <span class="method">GET</span> /health - Health check
                </div>
                <div class="endpoint">
                    <span class="method">GET</span> /config/{file} - Get configuration
                </div>
                
                <h2>Examples:</h2>
                <ul>
                    <li><a href="/health">/health</a> - Check server health</li>
                    <li><a href="/config/peanut.tsk">/config/peanut.tsk</a> - Get peanut configuration</li>
                </ul>
            </body>
            </html>
            """;
    }
    
    private static void handleConfigRequest(PrintWriter out, String filename, CommandLine globalCmd) {
        try {
            if (!Files.exists(Paths.get(filename))) {
                sendResponse(out, "404 Not Found", "application/json", 
                    "{\"error\":\"Configuration file not found: " + filename + "\"}");
                return;
            }
            
            TuskLangEnhanced parser = new TuskLangEnhanced();
            Map<String, Object> config = parser.parseFile(filename);
            
            // Convert to JSON
            com.fasterxml.jackson.databind.ObjectMapper mapper = new com.fasterxml.jackson.databind.ObjectMapper();
            String json = mapper.writeValueAsString(config);
            
            sendResponse(out, "200 OK", "application/json", json);
            
        } catch (Exception e) {
            sendResponse(out, "500 Internal Server Error", "application/json", 
                "{\"error\":\"" + e.getMessage() + "\"}");
        }
    }
    
    private static String generateCompiledContent(Map<String, Object> config, String originalFile) {
        StringBuilder sb = new StringBuilder();
        sb.append("# Compiled TuskLang Configuration\n");
        sb.append("# Original file: ").append(originalFile).append("\n");
        sb.append("# Compiled at: ").append(java.time.LocalDateTime.now()).append("\n");
        sb.append("# This file was automatically generated - do not edit manually\n\n");
        
        // Convert config back to TSK format
        sb.append(convertToTskFormat(config, 0));
        
        return sb.toString();
    }
    
    private static String generateOptimizedContent(Map<String, Object> config, String originalFile) {
        StringBuilder sb = new StringBuilder();
        sb.append("# Optimized TuskLang Configuration\n");
        sb.append("# Original file: ").append(originalFile).append("\n");
        sb.append("# Optimized at: ").append(java.time.LocalDateTime.now()).append("\n");
        sb.append("# This file was automatically optimized - do not edit manually\n\n");
        
        // Convert optimized config back to TSK format
        sb.append(convertToTskFormat(config, 0));
        
        return sb.toString();
    }
    
    private static Map<String, Object> optimizeConfiguration(Map<String, Object> config) {
        // Apply various optimizations
        Map<String, Object> optimized = new java.util.LinkedHashMap<>();
        
        for (Map.Entry<String, Object> entry : config.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();
            
            // Remove comments and empty values
            if (key.startsWith("#") || value == null) {
                continue;
            }
            
            // Optimize string values (remove unnecessary quotes)
            if (value instanceof String) {
                String str = (String) value;
                if (str.matches("^[a-zA-Z0-9_-]+$")) {
                    // Don't quote simple identifiers
                    optimized.put(key, str);
                } else {
                    optimized.put(key, value);
                }
            } else {
                optimized.put(key, value);
            }
        }
        
        return optimized;
    }
    
    private static String convertToTskFormat(Object value, int indent) {
        StringBuilder sb = new StringBuilder();
        String indentStr = "  ".repeat(indent);
        
        if (value instanceof Map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> map = (Map<String, Object>) value;
            
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                String key = entry.getKey();
                Object val = entry.getValue();
                
                if (val instanceof Map || val instanceof java.util.List) {
                    sb.append(indentStr).append(key).append(":\n");
                    sb.append(convertToTskFormat(val, indent + 1));
                } else {
                    sb.append(indentStr).append(key).append(": ");
                    if (val instanceof String) {
                        sb.append("\"").append(val).append("\"");
                    } else {
                        sb.append(val);
                    }
                    sb.append("\n");
                }
            }
        } else if (value instanceof java.util.List) {
            @SuppressWarnings("unchecked")
            java.util.List<Object> list = (java.util.List<Object>) value;
            
            for (Object item : list) {
                sb.append(indentStr).append("- ");
                if (item instanceof String) {
                    sb.append("\"").append(item).append("\"");
                } else {
                    sb.append(item);
                }
                sb.append("\n");
            }
        }
        
        return sb.toString();
    }
} 